book/src/dev/rfcs/XXXX-asynchronous-script-verification.md

```diff
@@ -0,0 +1,231 @@
+
+    // if outpoint is a hashmap key, remove the entry and send output on the channel
+    pub fn respond(&mut self, outpoint: OutPoint, output: TransparentOutput);
+
+    // extracts a list of new outputs from the block and checks them against the
+    // hashmap keys
+    pub fn check_block(&mut self, block: &Block);
+
+    // scans the hashmap and removes any entries with closed senders
+    pub fn prune(&mut self);
+}
+```
+
+The state service should maintain an `Arc<Mutex<PendingUtxos>>`, used as follows:
+
+1. In `Service::call(Request::AwaitUtxo(u))`, the service should:
+  - call `PendingUtxos::queue(u)` to get a future `f` to return to the caller;
+  spawn a task that does a sled lookup for `u`, calling `PendingUtxos::respond(u, output)` if present;
+  - check the in-memory storage for `u`, calling `PendingUtxos::respond(u, output)` if present;
+  - return `f` to the caller (it may already be ready).
+  The common case is that `u` references an old UTXO, so spawning the lookup
+  task first means that we don't wait to check in-memory storage for `u`
+  before starting the sled lookup.
+
+2. In `Service::call(Request::CommitBlock(block, ..))`, the service should:
+  - call `PendingUtxos::check_block(block.as_ref())`;
+  - do any other transactional checks before committing a block as normal.
+  Because the `AwaitUtxo` request is informational, there's no need to do
+  the transactional checks before matching against pending UTXO requests,
+  and doing so upfront potentially notifies other tasks earlier.
+
+3. In `Service::poll_ready()`, the service should call
+  `PendingUtxos::prune()` at least *some* of the time. This is required because
+  when a consumer uses a timeout layer, the cancelled requests should be
+  flushed from the queue to avoid a resource leak. However, doing this on every
+  call will result in us spending a bunch of time iterating over the hashmap.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+One drawback of this design is that we may have to wait on a lock. However,
+the critical section basically amounts to a hash lookup and a channel send,
+so I don't think that we're likely to run into problems with long contended
+periods, and it's unlikely that we would get a deadlock.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+High-level design rationale is inline with the design sketch. One low-level
+option would be to avoid encapsulating behavior in the `PendingUtxos` and
+just have an `Arc<Hashmap<..>>`, so that the lock only protects the hashmap
+lookup and not sending through the channel. But I think the current design is
+cleaner and the cost is probably not too large.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- We need to pick a timeout for UTXO lookup. This should be long enough to
+account for the fact that we may start verifying blocks before all of their
+ancestors are downloaded.
```

Comment:
Question: Are there two different kinds of timeout here?

1. If we are waiting to verify a block B, and we commit B's previous block, then we can be certain that we know the status of each UTXO in B. Each UTXO is either available now, or it will never become available.

Can we respond to all pending requests with a UTXO or error when B's parent is committed?
Should we do this?
Is it just an optimisation?

2. If we are waiting to verify a block C, but we have just started verifying from the genesis block, then it might be minutes or hours until all of C's UTXOs become available. Even if they do because available, we can't commit C until we commit all of its predecessors. In this case, we should timeout and reject C, then re-download it when we are closer to its height in the chain.

Should we timeout after some small multiple of the expected block interval?
Does the timeout here depend on C's (unverified) height?
Does the timeout depend on the re-download behaviour of the sync service?

Reply:
In case (1), I don't think that we have enough information to flush requests originating from block B with a failure, because the requests are for UTXOs by "outpoint"s (tx hash + output index), so there's not a way to trace them to a specific block as would be required.  We could add source tracking information, but I think that in this case it would be fine just to let them timeout.

In case (2), I think that it's unlikely that we'll want to start verifying C if the chain tip is very very very far behind C (e.g., all the way back to genesis).  If we did start verifying it, it would time out, but we don't remember any information about block verification failures, so there's no problem with just retrying.  The sync logic is *supposed* to restart syncing if it ever gets stuck.

I think that to start it would work to just set an unprincipled timeout (how about 120 seconds?) and then see how it works with the rest of the verification pipeline.  My guess is that we won't be timing out very often, since we'd presumably fail other checks first.

Reply:
When we're at the tip, blocks should arrive every ~70 seconds.

So I think the timeout should be greater than 70 seconds, but less than some small multiple of the block interval.

That way, we don't timeout on the same block multiple times, but we do get new blocks reasonably often.